src/bms-state.lisp

```diff
@@ -0,0 +1,70 @@
+(defun bms-state-loop () {
+    (var bms-vmin-limit-start (conf-get 'bms-vmin-limit-start))
+    (var bms-vmax-limit-start (conf-get 'bms-vmax-limit-start))
+    (var bms-tmax-limit-start (- (conf-get 'bms-t-limit-start) 3)) ; start 3 degrees before we start limiting
+    (var bms-cell-balance-start 0.5)
+    (var bms-discharge-tmax-limit-start (+ bms-tmax-limit-start 10)) ; set bms mosfet discharge limit to +10 deg c past cells
+    (var bms-config-update-time (systime))
+    (var bms-fault 0u32)
+
+    (var v-cell-support (>= (+ (first (sysinfo 'fw-ver)) (* (second (sysinfo 'fw-ver)) 0.01)) 6.06))
+
+    (var bms-fault-codes '(
+        (NONE . 0)
+        (BMS_COMM_TIMEOUT . 1)
+        (BMS_OVER_TEMP . 2)
+        (CELL_OVER_VOLTAGE . 3)
+        (CELL_UNDER_VOLTAGE . 4)
+        (CELL_OVER_TEMP . 5)
+        (CELL_UNDER_TEMP . 6)
+        (CELL_BALANCE . 7)
+    ))
+    (defun bms-set-fault (fault-code) {
+        (if (eq fault-code 'NONE) {
+            (setq bms-fault 0u32)
+        }{
+            (setq bms-fault (bitwise-or bms-fault (shl 1 (- (assoc bms-fault-codes fault-code) 1))))
+        })
+    })
+
+    (loopwhile t {
+        (if (and (<= (get-duty) 0.05)  (> (secs-since bms-config-update-time) 1.0)) { ; Only bother updating config values while board is idle
```

Comment:
I'm not too keen on the duty > 0.05 check, looks somewhat arbitrary. Stuff like this tends to strike back at some point. e.g., what if you set your board to flywheel and decide to finally fix your bms thresholds? And somehow the flywheel never goes under 0.05 duty (I have no clue if it actually does or not).

Wouldn't there per chance be some sort of a callback on the API to notify of config changes?

Reply:
> d->motor.duty_cycle > 0.05 is used in main.c for determining if board is should do HV/LV tiltback.

Yes. That's quite a different purpose for which it's used, I don't know how you draw the connection to here and use the same mechanism.

> If there's a callback on the API for config changes let me know.

Ah. I thought you'd be more familiar with the interface and it'd be easier for you to find.

I had a look and didn't find it.

> If someone was in flywheel mode and changed a config to fix bms thresholds (why you'd want to even change these from defaults I have no clue), the value would be updated as soon as 1 second after it's disengaged.

You didn't understand the example situation. I was trying to point out the values could have been needed before disengaging the flywheel.

I think you could just poll for the config changes every second and it'll be fine, though not optimal...

Reply:
poll every second